internal/venice-common/src/main/java/com/linkedin/venice/pubsub/listener/StoreChangeNotifier.java

```diff
@@ -0,0 +1,379 @@
+import com.linkedin.venice.pubsub.listener.StoreChangeTasks.StoreDeletedTask;
+import com.linkedin.venice.pubsub.listener.StoreChangeTasks.VersionAddedTask;
+import com.linkedin.venice.pubsub.listener.StoreChangeTasks.VersionDeletedTask;
+import com.linkedin.venice.utils.DaemonThreadFactory;
+import com.linkedin.venice.utils.LogContext;
+import it.unimi.dsi.fastutil.ints.IntOpenHashSet;
+import it.unimi.dsi.fastutil.ints.IntSet;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.annotation.concurrent.ThreadSafe;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * Thread-safe notifier for store metadata changes in PubSub clients.
+ *
+ * <p>This notifier handles store data change events (creation, deletion, version changes) and
+ * propagates them to registered PubSub adapters by executing their registered tasks asynchronously.
+ *
+ * <p><b>Key Features:</b>
+ * <ul>
+ *   <li>Event-specific tasks: Adapters register different tasks for different event types</li>
+ *   <li>Rich context: Tasks receive full Store objects and version numbers</li>
+ *   <li>Async execution: Tasks run in a thread pool to avoid blocking</li>
+ *   <li>Failure isolation: Exceptions in one task don't affect others</li>
+ *   <li>Minimal state: Only tracks version sets for change detection</li>
+ * </ul>
+ *
+ * <p><b>Usage Example:</b>
+ * <pre>{@code
+ * StoreChangeNotifier notifier = new StoreChangeNotifier(
+ *     4, // thread pool size
+ *     logContext,
+ *     VeniceComponent.SERVER
+ * );
+ *
+ * StoreChangeTasks tasks = StoreChangeTasks.builder()
+ *     .onVersionAdded((store, version) -> handleNewVersion(store, version))
+ *     .build();
+ *
+ * String taskId = notifier.registerTasks("MyAdapter", tasks);
+ *
+ * // Later, when closing
+ * notifier.unregisterTasks(taskId);
+ * notifier.close();
+ * }</pre>
+ *
+ * <p><b>Thread Safety:</b> All public methods are thread-safe and can be called concurrently.
+ */
+@ThreadSafe
+public class StoreChangeNotifier implements StoreDataChangedListener, AutoCloseable {
+  private static final Logger LOGGER = LogManager.getLogger(StoreChangeNotifier.class);
+  private static final int DEFAULT_THREAD_POOL_SIZE = 1;
+  private static final long SHUTDOWN_TIMEOUT_SECONDS = 30;
+
+  private final ExecutorService notificationExecutor;
+  private final ConcurrentHashMap<String, StoreChangeTasks> taskRegistry;
+  private final Map<String, IntSet> storeVersionSets;
+  private final AtomicInteger taskIdSuffixCounter;
+  private final AtomicBoolean closed;
+  private final VeniceComponent veniceComponent;
```

Comment:
Wondering why we need VeniceComponent here, will it be used for `StoreChangeTasks`?